beacon-chain/db/filesystem/data_column.go

```diff
@@ -0,0 +1,966 @@
+// The returned chanel is closed when the pruning is complete.
+func (dcs *DataColumnStorage) Store(dataColumnSidecars []blocks.VerifiedRODataColumn) error {
+	startTime := time.Now()
+
+	if len(dataColumnSidecars) == 0 {
+		return nil
+	}
+
+	// Check the number of columns is the one expected.
+	// While implementing this, we expect the number of columns won't change.
+	// If it does, we will need to create a new version of the data column sidecar file.
+	if params.BeaconConfig().NumberOfColumns != mandatoryNumberOfColumns {
+		return ErrWrongNumberOfColumns
+	}
+
+	highestEpoch := primitives.Epoch(0)
+	dataColumnSidecarsbyRoot := make(map[[fieldparams.RootLength]byte][]blocks.VerifiedRODataColumn)
+
+	// Group data column sidecars by root.
+	for _, dataColumnSidecar := range dataColumnSidecars {
+		// Check if the data column index is too large.
+		if dataColumnSidecar.ColumnIndex >= mandatoryNumberOfColumns {
+			return ErrDataColumnIndexTooLarge
+		}
+
+		// Group data column sidecars by root.
+		root := dataColumnSidecar.BlockRoot()
+		dataColumnSidecarsbyRoot[root] = append(dataColumnSidecarsbyRoot[root], dataColumnSidecar)
+	}
+
+	for root, dataColumnSidecars := range dataColumnSidecarsbyRoot {
+		// Safety check all data column sidecars for this root are from the same slot.
+		firstSlot := dataColumnSidecars[0].SignedBlockHeader.Header.Slot
+		for _, dataColumnSidecar := range dataColumnSidecars[1:] {
+			if dataColumnSidecar.SignedBlockHeader.Header.Slot != firstSlot {
+				return ErrDataColumnSidecarsFromDifferentSlots
+			}
+		}
+
+		// Set the highest epoch.
+		epoch := slots.ToEpoch(dataColumnSidecars[0].Slot())
+		highestEpoch = max(highestEpoch, epoch)
+
+		// Save data columns in the filesystem.
+		if err := dcs.saveFilesystem(root, epoch, dataColumnSidecars); err != nil {
+			return errors.Wrap(err, "save filesystem")
+		}
+
+		// Get all indices.
+		indices := make([]uint64, 0, len(dataColumnSidecars))
+		for _, dataColumnSidecar := range dataColumnSidecars {
+			indices = append(indices, dataColumnSidecar.ColumnIndex)
+		}
+
+		// Compute the data columns ident.
+		dataColumnsIdent := DataColumnsIdent{Root: root, Epoch: slots.ToEpoch(dataColumnSidecars[0].Slot()), Indices: indices}
+
+		// Set data columns in the cache.
+		if err := dcs.cache.set(dataColumnsIdent); err != nil {
+			return errors.Wrap(err, "cache set")
+		}
+
+		// Notify the data column feed.
+		dcs.DataColumnFeed.Send(dataColumnsIdent)
+	}
+
+	dataColumnSaveLatency.Observe(float64(time.Since(startTime).Milliseconds()))
+
+	return nil
+}
+
+// saveFilesystem saves data column sidecars into the database.
+// This function expects all data column sidecars to belong to the same block.
+func (dcs *DataColumnStorage) saveFilesystem(root [fieldparams.RootLength]byte, epoch primitives.Epoch, dataColumnSidecars []blocks.VerifiedRODataColumn) error {
+	// Compute the file path.
+	filePath := filePath(root, epoch)
+	exists, err := afero.Exists(dcs.fs, filePath)
```

Comment:
Wouldn't this check be racy? You could have multiple goroutines trying to call `saveDataColumnSidecarsNewFile`, and the way `os.Create` works, it will actually truncate the file when it is "created" the second time.

Reply:
Nice catch, thanks.

Fixed in https://github.com/prysmaticlabs/prysm/pull/15061/commits/bc017ed4930a56769546f03c6e0b4b9d2d07978c.